File: ios/RCTImageMarker/TextOptions.m
//
//  NSObject+TextOptions.m
//  react-native-image-marker
//
//  Created by Jimmydaddy on 2023/6/24.
//

#import "TextOptions.h"

@implementation TextOptions: NSObject

-(id)initWithDicOpts: (NSDictionary *) opts
{
    if (![[opts allKeys] containsObject:@"text"] || opts[@"text"] == nil) {
        @throw [NSException exceptionWithName:@"PARAMS_REQUIRED" reason:@"text is required" userInfo:nil];
    }

    NSDictionary* positionOpts = [opts objectForKey:@"positionOptions"]? opts[@"positionOptions"] : nil;
    if (positionOpts != nil) {
        _X = [RCTConvert CGFloat:positionOpts[@"X"]];
        _Y = [RCTConvert CGFloat:positionOpts[@"Y"]];
        _position = [positionOpts objectForKey: @"position"]? [RCTConvert MarkerPosition: positionOpts[@"position"]] : 0;
    } else {
        _X = 20.0f;
        _Y = 20.0f;
        _position = 0;
    }
    _text = opts[@"text"];
    _style = [[TextStyle alloc] initWithDicOpts: opts[@"style"]];
    return self;
}

@end

File: ios/RCTImageMarker/TextOptions.swift
//
//  TextOptions.swift
//  react-native-image-marker
//
//  Created by Jimmydaddy on 2023/6/24.
//

import Foundation
import UIKit
import React

class TextOptions: NSObject {
    var X: String?
    var Y: String?
    var edgeInset: String?
    var position: MarkerPositionEnum = .none
    var text: String
    var alpha: CGFloat = 1.0
    var style: TextStyle
    var layout: ImageMarkerWatermarkLayout?

    init(dicOpts opts: [AnyHashable: Any]) throws {
        guard let text = opts["text"] as? String else {
            throw NSError(domain: ErrorDomainEnum.PARAMS_REQUIRED.rawValue, code: 0, userInfo: [NSLocalizedDescriptionKey: "text is required"])
        }

        let positionOpts = opts["position"] as? [AnyHashable: Any]
        if let positionOpts {
            self.X = Utils.isNULL(positionOpts["X"]) ? nil : Utils.handleDynamicToString(v: positionOpts["X"])
            self.Y = Utils.isNULL(positionOpts["Y"]) ? nil : Utils.handleDynamicToString(v: positionOpts["Y"])
            self.edgeInset = Utils.isNULL(positionOpts["edgeInset"]) ? nil : Utils.handleDynamicToString(v: positionOpts["edgeInset"])
            self.position = positionOpts["position"] != nil ? RCTConvert.MarkerPosition(positionOpts["position"]) : .none
        }

        if let layoutOpts = opts["layout"] as? [AnyHashable: Any] {
            self.layout = try ImageMarkerWatermarkLayout(dicOpts: layoutOpts)
        }
        if layout?.isTile == true, positionOpts != nil {
            throw NSError(
                domain: ErrorDomainEnum.PARAMS_INVALID.rawValue,
                code: 0,
                userInfo: [NSLocalizedDescriptionKey: "layout cannot be combined with position"]
            )
        }

        self.text = text
        let styleOpts = opts["style"] as? [AnyHashable: Any] ?? [:]
        self.style = try TextStyle(dicOpts: styleOpts)
        self.alpha = try Utils.resolvedAlpha(opts["alpha"], fieldName: "text")
    }
}
